### PR TITLE
Add `py-exec` directive to execute python snippets

### PR DESCRIPTION
**Pull request by sizmailov**
Directive effectively runs `exec(<snippet code>)` and
renders highlighted code & output (if any)

Depends on `m.code` plugin

Features:
  - Allowed to created "contexts" to preserve state
    between snippets
  - code/stdout/stderr render can be optionally suppressed
  - top-level exceptions treated as errors
  - snippetes with expected exceptions must be
    explicitly annotated with `:raises: ExpectedExceptionType`
 
<details><summary>Render of  <code>test/py_exec/page.rst</code> :</summary>

image

</details>

### CODE REVIEW

File: plugins/m/test/test_py_exec.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+import re
+
+from . import PelicanPluginTestCase
+
+_css_colors_src = re.compile(r"""<span class="mh">#(?P<hex>[0-9a-f]{6})</span>""")
+_css_colors_dst = r"""<span class="mh">#\g<hex><span class="m-code-color" style="background-color: #\g<hex>;"></span></span>"""
+
+def _add_color_swatch(str):
+    return _css_colors_src.sub(_css_colors_dst, str)
```
Comment (mosra) [inline plugins/m/test/test_py_exec.py:34]: This and the `M_CODE_FILTERS` options below are not needed here, it was just testing code snippet filtering in the m.code plugin.

File: css/m-components.css [hunk 2]
```diff
@@ -634,6 +634,52 @@ figure.m-figure.m-fullwidth:after {
...
+}
+
+.m-code-figure > div.m-py-exec > pre.m-nopad, .m-console-figure > div.m-py-exec > pre.m-nopad {
+  padding-left: 0.875rem;
+  border-top-left-radius: 0;
+  border-bottom-right-radius: 0;
+  margin-top: 1rem;
+  margin-bottom: 0;
+  max-height: 20rem;
+}
```
Comment (mosra) [inline css/m-components.css:664]: Hmmm, this seems to be an exact copy of the block above?

**Comment by mosra** [comment]
Cool idea :+1: A while ago I was investigating doing something similar to [Sphinx Gallery](https://sphinx-gallery.github.io/) -- there it's able to not only print the output, but also extract matplotlib graphs or show images. Such functionality would be useful for easier creation of various tutorial pages such as [this one in Habitat](https://aihabitat.org/docs/habitat-sim/lighting-setups.html), and this PR is going in that direction. Nice.

What I'm not sure about, however, is the labels placed on top of the code. It adds quite a bit of CSS just for this plugin alone, and (this is probably a very personal opinion but) I think it's rather gimmicky: 

- in most cases I'd want to have stderr and stdout interleaved together, so the stderr/stdout markings don't make sense to me (imagine getting output from `make` and warnings from `gcc` in separate terminals, not useful)
- if something throws an exception, I would just show the output error message, without duplicating the info to a label
- looking at the screenshot, the labels always grab my attention and i can't stay focused on the main content with those

With the improved ANSI color output from #151 I think this could be solved for example by marking standard error output in red. Alternatively, image figures support color classes (see [an example here, scroll down](https://mcss.mosra.cz/css/components/test/#figures)), which could be potentially extended to code figures as well.

**Comment by sizmailov** [comment]
matplotlib figures would be great to have but I've put it aside for a while. 

Yeah, HTML/CSS is weakest part of PR, no doubt. I agree, labels should be more pale if any. 
We can easily mix stdout/stderr but we will not be able to separate them later for colorizing (at least easialy) 

Would it make sense to 
 - withdraw labels and corresponding css from PR 
 - mix stdout/stderr in one stream
? 

Once we have basic version of plugin we can work on HTML/CSS.

**Comment by mosra** [comment]
For the stderr/stdout colorization I was thinking if this could be done on the "executor" side somehow? Like, intercepting both stderr and stdout, colorizing them, and then finally passing to the output. But I don't think this is critical.

Yes, I would remove the CSS and label stuff, at least for now. With that gone the plugin is nicely self-contained, which means it could be integrated in projects even without having to use a certain fork of m.css. Like I said above, I have something like this on my tasklist (plus further extension/conversion of the rendered page into a ipynb file), but I don't know yet which direction to go, so if it's okay, can we keep it in a branch for now? :)

By the way, how's the scoping done with `exec()`? Like, if I have two files and each of them imports, defines and executes something else, wouldn't the two clash together (suddenly functions defined in one being executed in the other etc.)?

**Comment by sizmailov** [comment]
Terminal-like capture of stdout/stderr is possible, it's only matter of time to implement :)

I'll remove label-related stuff from PR. 

Scoping done by simply storing number of `globals` dicts, one per scope. It should be isolated enough for simple cases, but some interference is possible, e.g. python modules are imported just once (i.e. module initialization code is executed once per interpreter run). Two `.rst` files may interfere through this global dicts, but I guess the order of processing `.rst` files is not defined, so it's not desired. Currently it's user's responsibility to not mess things up. We can do some checks/auto cleanup, but it's still not 100% proof.

**Comment by sizmailov** [comment]
Labels removed, stdout/stderr are combined, rebased on top of master. 

<details><summary>preview</summary>

image

</details>

**Comment by mosra** [comment]
I have reconsidered my position on this PR -- since this is a rather niche feature and the functionality is quite standalone, I think it could live on its own outside of the m.css repository, maintained by a third party instead.

Of course if there would be any future modifications needed on m.css side, I'm happy to accomodate. But having this builtin, including the additions to the global CSS that majority of users wouldn't ever use, would only add a maintenance burden for me. Especially given that I didn't find any use for this feature so far. I hope that's understandable.

**Comment by sizmailov** [comment]
I see your point. 

FYI: I think at the time the features I was interested were:
 - enforced sync of doc snippets with current library code
 - reduced burden on writing "examples" doc section (... and "gallery")

**Comment by mosra** [comment]
No problem. Sorry for not having more time for m.css when you were using it.

**Pull request closed without merging**